### PR TITLE
[BugFix] [Performance] fix dreamer v3 training semantics

### PR DESCRIPTION
**Pull request by gtnv**
### TLDR

working on this part:
> that the new algos that we have provided all have nice looking learning curves (especially dreamer 3 on some non trivial task)

This pr fixes DreamerV3's real-policy, reset, reconstruction, and actor-loss paths. The real policy now uses each observation and shares the trained world model.

Connected to/based on: https://github.com/pytorch/rl/pull/4118. It does not duplicate that pr's device, CPU-replay, or history changes.

<details><summary>Testing + visuals</summary>
<p>

### DMC Walker Stand

```
NVIDIA H100 PCIe
646,819 unique optimizer-owned parameters
24D observation, 6D continuous action
seed 0, 3,000 real steps and 3,000 updates
B=8, T=16, imagination horizon 15
3 evaluation episodes per point
```

image

- Reconstruction loss: 22.420 to 2.958 (-86.8%).
- Reward loss: 5.387 to 0.576 (-89.3%).
- Raw return first-four median: 105.19; last-four median: 128.70 (+22.3%); final: 212.15.

these are raw, unsmoothed points from a complete exit-0 fixed-worktree run. the curve log predates the semantic commit and has no embedded SHA, and one 3k-step seed is not benchmark evidence.

### Functional controls

```
baseline 6e18e35b
fixed semantics 8fdf0b5e
same seed and paired observations
```

image

- Observation gradient norm: 0 to 0.003504; paired-observation action delta: 0 to 0.004378.
- Shared world-model parameter tensors: 0 to 48; next belief: absent to present.
- The baseline deterministic caller exits 1 on `TanhNormal.mode`; the fixed caller exits 0.
- The exact CI lint command passes locally; 268 Dreamer consumer tests pass.

### Reproduction

```bash
/usr/bin/time -v python sota-implementations/dreamer_v3/dreamer_v3.py \
  --config-name=config_dmc_walker env.task=stand \
  optimization.device=cuda optimization.use_reinforce=false \
  collector.total_frames=3000 collector.frames_per_batch=250 \
  replay_buffer.buffer_size=3000 replay_buffer.batch_size=8 \
  replay_buffer.seq_len=16 replay_buffer.warmup_factor=1 \
  optimization.train_ratio=null optimization.updates_per_batch=250 \
  logger.eval_every=250 logger.eval_episodes=3 \
  logger.metrics_json=/tmp/walker_stand.json \
  2>&1 | tee /tmp/walker_stand.log
```

```python
import re
from pathlib import Path

import matplotlib.pyplot as plt

pattern = re.compile(
    r"\[env_step=\s*(\d+)\]\s+eval_reward=([+-]?\d+(?:\.\d+)?)\s+"
    r"kl=\S+\s+reco=(\d+(?:\.\d+)?)\s+reward=(\d+(?:\.\d+)?)"
)
points = [tuple(map(float, match.groups())) for match in pattern.finditer(Path("/tmp/walker_stand.log").read_text())]
steps, returns, reconstruction, reward = zip(*points)
plt.style.use("dark_background")
figure, axes = plt.subplots(1, 2)
axes[0].plot(steps, returns, marker="o")
axes[1].plot(steps, reconstruction, marker="o", label="reconstruction")
axes[1].plot(steps, reward, marker="o", label="reward")
axes[1].set_yscale("log")
figure.savefig("dreamer_v3_h100_learning_dark.png")
```

```bash
pre-commit run --all-files
pytest test/modules/test_dreamer_components.py test/objectives/test_dreamer_v3.py -q
```

```python
real_actor(real_input)
gradient = torch.autograd.grad(real_input["loc"].sum(), observation)[0]
assert gradient.abs().sum() > 0
assert ("next", "belief") in real_input.keys(include_nested=True)
```

</p>
</details>

ai disclosure: in this pr i used an agent for helping me format and write the pr description and help with env testing

cc @vmoens @theap06

### CODE REVIEW

**Review by vmoens — APPROVED** [review]
LGTM thanks
I have a PR to make a Minecraft wrapper (#4107) that we should use to reprod the paper results at some point if compute is available

**Comment by theap06** [comment]
> LGTM thanks I have a PR to make a Minecraft wrapper (#4107) that we should use to reprod the paper results at some point if compute is available

@vmoens been meaning to reprod the minecraft results for the dreamer paper. Could I reprod the results with the current pr that you put up?

**Comment by vmoens** [comment]
Of course 
Maybe let's give a quick review to the Minecraft PR first?
And let's check the run speed on collectors / async envs?
I suspect resets will take ages, making async envs a necessary backend for parallel execution

**Comment by gtnv** [comment]
@theap06 lmk if u need my help. will review 4107
